src/shared.rs

```diff
@@ -0,0 +1,211 @@
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::ops::Deref;
+use std::sync::RwLock;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use {Future, Poll, Async};
+use task::{self, Task};
+
+#[derive(Debug)]
+pub struct SharedItem<T> {
+    item: Arc<T>,
+}
+
+impl<T> Clone for SharedItem<T> {
+    fn clone(&self) -> Self {
+        SharedItem { item: self.item.clone() }
+    }
+}
+
+#[derive(Debug)]
+pub struct SharedError<E> {
+    error: Arc<E>,
+}
+
+impl<T> Clone for SharedError<T> {
+    fn clone(&self) -> Self {
+        SharedError { error: self.error.clone() }
+    }
+}
+
+impl<T> SharedItem<T> {
+    fn new(item: T) -> Self {
+        SharedItem { item: Arc::new(item) }
+    }
+}
+
+impl<E> SharedError<E> {
+    fn new(error: E) -> Self {
+        SharedError { error: Arc::new(error) }
+    }
+}
+
+struct SyncedInner<F>
+    where F: Future
+{
+    original_future: F,
+    result: Option<Result<Async<SharedItem<F::Item>>, SharedError<F::Error>>>,
+    tasks_receiver: Mutex<Receiver<Task>>,
```

Comment:
It looks like the blocking functionality of this `Mutex` is never used, so perhaps this could be `Lock` (in this repo)